api/external_auth.proto

```diff
@@ -0,0 +1,152 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2;
+
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with OK or NOT OK.
+service Authorization {
+   rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+    // The name of the tenant for multi-tenant support.
+    string tenant_name = 1;
+
+    // The request attributes.
+    AttributeContext attributes = 2;
+}
+
+message CheckResponse {
+    // Status OK allows the request. Any other response indicates the request should be denied.
+    google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// The fields follow the Istio attribute vocabulary.
+// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+// The field type becomes the attribute type, and the field path becomes the attribute name.
+// For example, the attribute `source.ip` maps to field `AttributeContext.source.ip`.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The IP address of the peer.
+    string ip = 1;
+
+    // The network port of the peer.
+    int64 port = 2;
+
+    // The canonical service name of the peer.
+    string service = 3;
+
+    // The labels associated with the peer.
+    // These could be pod labels for Kubernetes or tags for VMs.
+    // The source of the labels could be an X.509 certificate or other configuration.
```

Comment:
Do we need to explain the source of labels? If so, we need to have more details here.

Reply:
Let's drop these until we have a concrete proposal for how to populate them.

Reply:
We can drop it for now.

Reply:
These are pod lablels like `app`, `version` and `environment` on which an operator can write a policy. We already use them in Istio. This should be part of the static filter configuration, which we have not addressed yet.

These labels are obtained from the environment. The mechanics of how labels are obtained should not block inclusion of the field, those proposals will follow.

Reply:
If it's not currently implemented in Envoy, please just drop it for now. We can add additional fields later. This is related to the Node/Locality concept also which needs to be reconciled IMO.